panic when using URL without credentials

```
$ protodep up     
[INFO] force update = false
[INFO] current dir is = /Users/mhelmich/REDACTED
[INFO] home dir is = /Users/mhelmich
[INFO] identity file = /Users/mhelmich/.ssh/id_rsa
[INFO] ssh port = 22
[INFO] found rewrite in gitconfig for 'gitlab.mittwald.it/REDACTED' ...
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/mittwald/protodep/logger.CensorHttpsPassword(0xc0000d05a0, 0x55, 0x22, 0xc00018615a)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/logger/logger.go:36 +0x1eb
github.com/mittwald/protodep/service.(*SyncImpl).getAuthProvider(0xc000172000, 0xc00018c240, 0x22, 0xc000180200, 0xc000178240, 0xc0001860f0, 0x45, 0x171cf20, 0xc000170000, 0x0, ...)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/service/sync.go:116 +0x1bc
github.com/mittwald/protodep/service.(*SyncImpl).getNewDeps(0xc000172000, 0xc00016e060, 0xc000164240, 0x3c, 0x3c, 0x1612ae0, 0xc000145d01)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/service/sync.go:170 +0x1b8
github.com/mittwald/protodep/service.(*SyncImpl).Resolve(0xc000172000, 0xc000172000, 0x20, 0xc000145d48)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/service/sync.go:59 +0x152
github.com/mittwald/protodep/cmd.glob..func1(0x1b35fc0, 0x1b58630, 0x0, 0x0, 0x0, 0x0)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/cmd/up.go:47 +0x44d
github.com/spf13/cobra.(*Command).execute(0x1b35fc0, 0x1b58630, 0x0, 0x0, 0x1b35fc0, 0x1b58630)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:826 +0x465
github.com/spf13/cobra.(*Command).ExecuteC(0x1b35d40, 0x100796f, 0xc000098058, 0x0)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/mhelmich/Git/Go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/mittwald/protodep/cmd.Execute()
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/cmd/root.go:37 +0x32
main.main()
	/Users/mhelmich/Git/Go/pkg/mod/github.com/mittwald/protodep@v0.1.20/main.go:6 +0x20
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,6 +16,7 @@ env:
     - GO_VERSION=1.12
     - secure: "SB7DKQgTBoqU+Jfc83y4n/H2NES/PrGD+7UCYy69lXyK354WKWTxtvHrEkecga0UiJAI9v1diwZiu5YEiJ0nLuk5vh9Jq9I47n7Y5sTRECn3h2UhqnuoVTGeULOKtAuRHlAZ+JG1eC9MsqG7SpENnS7HMJepoNfZZ7WrSPNVOE5GFR7ErdhR/tBahWPbXuAbtjhKdS6/LEsoxWf3aJAF98UEbEOYR8kXkUcXtdF5C6UX4R+X+5C+mU/JHTHK9dEpJehz2XSSAxsTY9ORvKPuL+JF2M2SoC2yTntZ7TBBiWsuwqszG4f33qk4p8H87tZjoHJiJpnN4hJBw8hL18+SrFM7adU1QcmKK+/iNMU2HO5KilAPleFKXhUX356sGvk8pXoqZEGvAAtzNr+Fwqwtpe+Df+fYzimHp2a0wqaIq6kf66CGlDLv6hKy10nh36Y2Wh3KHEywrbJEurXEOomtvy5p4o88wpv5PD5Q1mZZP2HxCdzaKAbyk/4UJ0BElbVwkUW38s5McWGUZxCSlURc9xrRoCWo2Kx4mn+29GEer8JNIzFzR8k6Slh+K1tGrzuUiG0gl7dt0UY+Ij+NbBSTnMDg3A3XnAOajolYdqQstVlFxhXJ+7MgCyk24Ksm2uAv6Z1n94FgPWBP249Za/WORwkzyILlD8CZHULtn9qMB74="
 script:
+  - go test ./...
   - curl -sL https://git.io/goreleaser | bash -s -- --snapshot --skip-publish --rm-dist
 
 deploy:

--- a/logger/logger.go
+++ b/logger/logger.go
@@ -2,7 +2,7 @@ package logger
 
 import (
 	"fmt"
-	"strings"
+	"net/url"
 	"time"
 
 	"github.com/briandowns/spinner"
@@ -26,17 +26,20 @@ func (s *spinnerWrapper) Finish() {
 	fmt.Print("\n")
 }
 
-func CensorHttpsPassword(url string) string {
-	path := strings.Split(url, "@")
-
-	if len(path) == 1 {
-		return url
+func CensorHttpsPassword(repoURL string) string {
+	parsed, err := url.Parse(repoURL)
+	if err != nil {
+		return fmt.Sprintf("<invalid url: %s>", err.Error())
 	}
-	cred := strings.Split(path[0], ":")
-	cred[2] = "xxxxxx"
-	compCred := strings.Join(cred, ":")
 
-	return compCred + "@" + path[1]
+	if parsed.User != nil {
+		_, pwdSet := parsed.User.Password()
+		if pwdSet {
+			parsed.User = url.UserPassword(parsed.User.Username(), "REDACTED")
+		}
+	}
+
+	return parsed.String()
 }
 
 func InfoWithSpinner(format string, a ...interface{}) *spinnerWrapper {

--- a/logger/logger_test.go
+++ b/logger/logger_test.go
@@ -1,0 +1,36 @@
+package logger
+
+import "testing"
+
+func TestCensorHttpsPasswordRemovesPasswordFromHTTPSURLWithPassword(t *testing.T) {
+	rawURL := "https://user:password@github.com/foo/bar.git"
+	censoredURL := CensorHttpsPassword(rawURL)
+
+	expectedCensoredURL := "https://user:REDACTED@github.com/foo/bar.git"
+
+	if censoredURL != expectedCensoredURL {
+		t.Error("incorrectly censored URL", "expected", expectedCensoredURL, "actual", censoredURL)
+	}
+}
+
+func TestCensorHttpsPasswordKeepsHTTPSURLWithoutPassword(t *testing.T) {
+	rawURL := "https://user@github.com/foo/bar.git"
+	censoredURL := CensorHttpsPassword(rawURL)
+
+	expectedCensoredURL := "https://user@github.com/foo/bar.git"
+
+	if censoredURL != expectedCensoredURL {
+		t.Error("incorrectly censored URL", "expected", expectedCensoredURL, "actual", censoredURL)
+	}
+}
+
+func TestCensorHttpsPasswordKeepsSSHURLs(t *testing.T) {
+	rawURL := "ssh://git@github.com/foo/bar.git"
+	censoredURL := CensorHttpsPassword(rawURL)
+
+	expectedCensoredURL := "ssh://git@github.com/foo/bar.git"
+
+	if censoredURL != expectedCensoredURL {
+		t.Error("incorrectly censored URL", "expected", expectedCensoredURL, "actual", censoredURL)
+	}
+}

